File: src/test/scala/com/exasol/cloudetl/kafka/deserialization/FieldParserTest.scala
package com.exasol.cloudetl.kafka.deserialization

import org.scalatest.funsuite.AnyFunSuite

class FieldParserTest extends AnyFunSuite {

  test("must parse list of field specs to correct fields") {
    assert(FieldParser.get(Seq("key", "value.*")) === Seq(RecordKey, RecordValueFields))
  }

  test("must parse field spec string string to correct fields") {
    assert(FieldParser.get("key, value.*") === Seq(RecordKey, RecordValueFields))
  }

  test("must parse to correct fields with timestamp and select") {
    val fields = Seq("key.field1", "timestamp", "value.*", "value.field2", "value.Field_3")
    val expectedFields =
      Seq(
        RecordKeyField("field1"),
        TimestampField,
        RecordValueFields,
        RecordValueField("field2"),
        RecordValueField("Field_3")
      )
    assert(FieldParser.get(fields) === expectedFields)
  }

  test("must parse to correct fields for names matching [A-Za-z_][A-Za-z0-9_]") {
    val fields = "key._key1, value.FirstValue, value.secondValue, value.value_3"
    val expectedFields =
      Seq[GlobalFieldSpecification](
        RecordKeyField("_key1"),
        RecordValueField("FirstValue"),
        RecordValueField("secondValue"),
        RecordValueField("value_3")
      )
    assert(FieldParser.get(fields) === expectedFields)
  }
}

File: src/test/java/com/exasol/cloudetl/kafka/deserialization/FieldParserTest.java
package com.exasol.cloudetl.kafka.deserialization;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsString;
import static org.junit.jupiter.api.Assertions.*;

import java.util.List;

import org.junit.jupiter.api.Test;

class FieldParserTest {
    @Test
    void parsesListOfFieldSpecsToCorrectFields() {
        assertEquals(List.of(RecordKey.INSTANCE, RecordValueFields.INSTANCE), FieldParser.get(List.of("key", "value.*")));
    }

    @Test
    void parsesFieldSpecStringToCorrectFields() {
        assertEquals(List.of(RecordKey.INSTANCE, RecordValueFields.INSTANCE), FieldParser.get("key, value.*"));
    }

    @Test
    void parsesToCorrectFieldsWithTimestampAndSelect() {
        final var fields = List.of("key.field1", "timestamp", "value.*", "value.field2", "value.Field_3");

        assertEquals(List.of(new RecordKeyField("field1"), TimestampField.INSTANCE, RecordValueFields.INSTANCE,
                new RecordValueField("field2"), new RecordValueField("Field_3")), FieldParser.get(fields));
    }

    @Test
    void parsesToCorrectFieldsForSupportedNames() {
        final String fields = "key._key1, value.FirstValue, value.secondValue, value.value_3";

        assertEquals(List.of(new RecordKeyField("_key1"), new RecordValueField("FirstValue"),
                new RecordValueField("secondValue"), new RecordValueField("value_3")), FieldParser.get(fields));
    }

    @Test
    void failsForInvalidFieldReferences() {
        final Exception thrown = assertThrows(Exception.class, () -> FieldParser.get(List.of("metadata.partition")));

        assertAll(() -> assertThat(thrown.getMessage(), containsString("E-KCE-14")),
                () -> assertThat(thrown.getMessage(), containsString("does not have the correct format")));
    }
}
